### ISSUE TITLE
Job.schedule() randomly sometimes not rescheduling

### ISSUE THREAD

**Issue by jukzi** [eclipse-platform/eclipse.platform#160]
This is a follow up from https://github.com/eclipse-platform/eclipse.platform/issues/148
The test org.eclipse.core.tests.runtime.jobs.Bug_574883.testJoinLambdaOften() sometimes still fail - but not because the join() is wrong but because the schedule() sometimes just does not reschedules the job accordingly.

I will provide a junit to show that.

**Comment by jukzi** [comment]
can not reproduce anymore

**Comment by basilevs** [comment]
I've prepared a reliable reproducer:

```
	@Test
	void doNotStopAfterSchedule() throws InterruptedException {
		Job job = BUSY_LOOP;
		for (int i = 0; i< 100; i++) {
			job.schedule();
			assertSoon(SCHEDULING_TOLERANCE, "Job have to start after schedule(). Iteration " + i, () -> job.getState() == Job.RUNNING);
			Thread.sleep(i);
			job.cancel();
			job.schedule();
			assertNever(SCHEDULING_TOLERANCE, "Infinite job can't stop after schedule(). Iteration " + i, () -> job.getState() == Job.NONE);
		}
	}
```

[JobTest.java](https://github.com/user-attachments/files/23093558/JobTest.java)

To reproduce, import [reproducer_160.zip](https://github.com/user-attachments/files/23093499/reproducer_160.zip) into PDE (works on any Target Platform from 2024-09 to 2025-12, probably more) and run launch configuration "JobTest".

Fails every time on:
> Eclipse IDE for Eclipse Committers (includes Incubating components)
> Version: 2025-12 M1 (4.38.0 M1)
> Build id: 20251010-0159
> OS: Mac OS X, v.12.7.6, aarch64 / cocoa
> Java vendor: Oracle Corporation
> Java runtime version: 25+36-3489
> Java version: 25

@jukzi  please reopen.

**Comment by basilevs** [comment]
Strange, this should already be detected by Bug_550738 regression test. I'll prepare a PR with an enhanced test.

**Comment by laeubi** [comment]
> I've prepared a reliable reproducer:

You should maybe explain a bit more *what* exactly it reproduces and where it fails

**Comment by basilevs** [comment]
> > I've prepared a reliable reproducer:
> 
> You should maybe explain a bit more _what_ exactly it reproduces and where it fails

This is effectively a regression of [Bug 550738](https://bugs.eclipse.org/bugs/show_bug.cgi?id=550738).

`Job.schedule()` is occasionally not effective (leaves Job in NONE state) when called soon after `Job.cancel()`

**Comment by iloveeclipse** [comment]
Not sure test is showing full picture, it seem to be a test for job.getState() output, not for the job scheduling.

Adding "printlns" to the job execution method shows that the job is scheduled properly and started even after the test failure.

```
package com.example.job_cancel_reproducer;

import static java.lang.System.currentTimeMillis;
import static org.junit.jupiter.api.Assertions.fail;

import java.time.Duration;
import java.util.function.BooleanSupplier;

import org.eclipse.core.runtime.CoreException;
import org.eclipse.core.runtime.ICoreRunnable;
import org.eclipse.core.runtime.IProgressMonitor;
import org.eclipse.core.runtime.jobs.IJobChangeEvent;
import org.eclipse.core.runtime.jobs.Job;
import org.eclipse.core.runtime.jobs.JobChangeAdapter;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class JobTest {
	private static final Duration SCHEDULING_TOLERANCE = Duration.ofMillis(500);
	
	private static final Job BUSY_LOOP = Job.create("Busy loop", new ICoreRunnable() {
		static int count;
		@Override
		public void run(IProgressMonitor monitor) throws CoreException {
			System.out.println("Job started " + (++count));
			while (!monitor.isCanceled()) {
				Thread.yield();
			}
		}
	});
	
	static {
		BUSY_LOOP.setPriority(Job.INTERACTIVE);
	}
	

	@BeforeEach
	@AfterEach
	public void stop() throws InterruptedException {
		BUSY_LOOP.cancel();
		BUSY_LOOP.join();
	}
	
	@Test
	void doNotStopAfterSchedule() throws InterruptedException {
		Job job = BUSY_LOOP;
		job.addJobChangeListener(new JobChangeAdapter() {
			@Override
			public void scheduled(IJobChangeEvent event) {
				System.out.println(" Job scheduled");
			}
		});
		for (int i = 0; i< 100; i++) {
			job.schedule();
			assertSoon(SCHEDULING_TOLERANCE, "Job have to start after schedule(). Iteration " + i, () -> job.getState() == Job.RUNNING);
			Thread.sleep(i);
			job.cancel();
			System.out.println(i + " State after cancel: " + job.getState());
			job.schedule();
			System.out.println(i + " State after schedule: " + job.getState());
			assertNever(SCHEDULING_TOLERANCE, "Infinite job can't stop after schedule(). Iteration " + i, () -> {
				int state = job.getState();
				if(Job.RUNNING != state) {
					System.out.println("State after cancel and schedule: " + state);
				}
				return state == Job.NONE;
			});
		}
	}

	private void assertSoon(Duration timeout, String message, BooleanSupplier condition) {
		long stop = currentTimeMillis() + timeout.toMillis();
		while (currentTimeMillis() < stop) {
			Thread.yield();
			if (condition.getAsBoolean()) {
				return;
			}
		}
		fail(message);
	}
	
	
	private void assertNever(Duration timeout, String message, BooleanSupplier condition) {
		long stop = currentTimeMillis() + timeout.toMillis();
		while (currentTimeMillis() < stop) {
			Thread.yield();
			if (condition.getAsBoolean()) {
				fail(message);
			}
		}
	}

}
```

**Comment by iloveeclipse** [comment]
Example output for me:
```
 Job scheduled
0 State after cancel: 4
 Job scheduled
0 State after schedule: 4
Job started 1
1 State after cancel: 4
1 State after schedule: 4
 Job scheduled
State after cancel and schedule: 0
Job started 2
```

**Comment by basilevs** [comment]
> Adding "printlns" to the job execution method shows that the job is scheduled properly and started even after the test failure.

You can't add println in these tests - they add a strong synchronization. I can adapt the test to use other means of checking if the job is still running. For example, job could reset a flag on every iteration and. we could check externally whether that flag is still active.

**Comment by iloveeclipse** [comment]
> For example, job could reset a flag on every iteration and. we could check externally whether that flag is still active.

Yes, please, but I would prefer counter. Flag has two states only, it will be not clear which execution changed that.

**Comment by basilevs** [comment]
> prefer counter.

The job does indeed seem to start.

The check:
```java
public boolean isRunning() {
	long first = iterationCount.get();
	Instant stop = now().plus(SCHEDULING_TOLERANCE);
	while (now().isBefore(stop)) {
		if (first != iterationCount.get()) {
			return true;
		}
		Thread.yield();
	}
	return false;
}
```
always returns `true`.

[JobTest.java](https://github.com/user-attachments/files/23095885/JobTest.java)

**Comment by basilevs** [comment]
RCPTT relied on `Job.NONE` state to detect that job is done and gone. Now if user code starts the job, RCPTT may miss the fact.
1. Should I create a separate issue to track `getState()`?
2. Is there a reliable way to ensure that UI thread has not scheduled execution yet?

**Comment by iloveeclipse** [comment]
I'm out of PC the rest of the day, but answers are
1) probably no, API from getState does not give any guarantees
2) Yes. Typically scheduling async task with a flag or check can be used

**Comment by basilevs** [comment]
> 2\. Yes. Typically scheduling async task with a flag or check can be used

My original test demonstrates, that even when performed in the same thread `getState()` is not reliable. What should such task query?

**Comment by basilevs** [comment]
3. What is the expected behavior of `join()`? Following test fails:

```
	@Test
	void doNotJoinAfterSchedule() throws InterruptedException {
		BusyLoop job = new BusyLoop();
		try {
			for (int i = 0; i< 100; i++) {
				job.schedule();
				assertTrue(job.isRunning());
				Thread.sleep(i);
				job.cancel();
				job.schedule();
				assertFalse(job.join(SCHEDULING_TOLERANCE.toMillis(), null)); // throws
			}
		} finally {
			job.cancel();
			job.join();
		}
	}
```

It's javadoc says:

> If this method is called on a job that reschedules itself from within the run method, the join will return at the end of the first execution. In other words, join will return the first time this job exits the RUNNING state, or as soon as this job enters the NONE state.

This seems to be predicated on job rescheduling itself, but it probably applicable to `cancel(); schedule();` sequence from outside of the job `run()` method.

**Comment by basilevs** [comment]
I've restated the problem in terms of `Job.join()` in #2210.

**Comment by jukzi** [comment]
Hi @basilevs, i have only run over your comments and could not see the direct violation of any contract. I even doubt that it may be possible to make an implementation that fullfills your assertions. However if you can somehow improve the implementation to fullfill any contract without touching the existing contracts/tests, then i am sure it would be accept. I myself have no time to dig into it and wish you all the best.

**Comment by basilevs** [comment]
> Hi [@basilevs](https://github.com/basilevs), i have only run over your comments and could not see the direct violation of any contract. I even doubt that it may be possible to make an implementation that fullfills your assertions. However if you can somehow improve the implementation to fullfill any contract without touching the existing contracts/tests, then i am sure it would be accept. I myself have no time to dig into it and wish you all the best.

In #2210 contracts are not violated either (there is a bad wording in Javadoc, that makes contracts seem stronger then they are), but it illustrates that `Job.join()` is basically useless.

I think this can be closed again, as the problem as originally stated is not reproducible - I was confused by `getState()` results.
